File: Bots/Python/Skills/CodeFirst/WaterfallSkillBot/dialogs/cards/card_dialog.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

import json

from botbuilder.core import MessageFactory
from botbuilder.dialogs import (
    ComponentDialog,
    DialogTurnResult,
    DialogTurnStatus,
    WaterfallDialog,
    WaterfallStepContext,
    Choice,
    ListStyle,
)
from botbuilder.dialogs.prompts import (
    ChoicePrompt,
    PromptOptions,
    PromptValidatorContext,
)
from botbuilder.schema import (
    Activity,
    ActivityTypes,
    EndOfConversationCodes,
    InputHints,
)
from config import DefaultConfig
from .card_options import CardOptions
from .channel_supported_cards import ChannelSupportedCards
from .card_sample_helper import CardSampleHelper


CORGI_ON_CAROUSEL_VIDEO = "https://www.youtube.com/watch?v=LvqzubPZjHE"
MIND_BLOWN_GIF = (
    "https://media3.giphy.com/media/xT0xeJpnrWC4XWblEk/giphy.gif?"
    "cid=ecf05e47mye7k75sup6tcmadoom8p1q8u03a7g2p3f76upp9&rid=giphy.gif"
)
MUSIC_API = "api/music"
TEAMS_LOGO_FILE_NAME = "teams-logo.png"


class CardDialog(ComponentDialog):
    def __init__(self, configuration: DefaultConfig):
        super().__init__(CardDialog.__name__)
        self.configuration = configuration

        self.add_dialog(ChoicePrompt(ChoicePrompt.__name__, self.card_prompt_validator))
        self.add_dialog(
            WaterfallDialog(
                WaterfallDialog.__name__,
                [self.select_card_step, self.display_card_step],
            )
        )

        self.initial_dialog_id = WaterfallDialog.__name__

    async def select_card_step(self, step_context: WaterfallStepContext):
        # Create the PromptOptions from the skill configuration which contain the list of configured skills.
        message_text = "What card do you want?"
        reprompt_message_text = "This message will be created in the validation code"

        options = PromptOptions(
            prompt=MessageFactory.text(
                message_text, message_text, InputHints.expecting_input
            ),
            retry_prompt=MessageFactory.text(
                reprompt_message_text, reprompt_message_text, InputHints.expecting_input
            ),
            choices=[Choice(card.value) for card in CardOptions],
            style=ListStyle.list_style,
        )

        return await step_context.prompt(ChoicePrompt.__name__, options)

    async def display_card_step(self, step_context: WaterfallStepContext):
        card_type = CardOptions(step_context.result.value)

        if ChannelSupportedCards.is_card_supported(
            step_context.context.activity.channel_id, card_type
        ):
            if card_type == CardOptions.adaptive_card_bot_action:
                await step_context.context.send_activity(
                    MessageFactory.attachment(
                        CardSampleHelper.create_adaptive_card_bot_action()
                    )
                )

            elif card_type == CardOptions.adaptive_card_task_module:
                await step_context.context.send_activity(
                    MessageFactory.attachment(
                        CardSampleHelper.create_adaptive_card_task_module()
                    )
                )

            elif card_type == CardOptions.adaptive_card_sumbit_action:
                await step_context.context.send_activity(
                    MessageFactory.attachment(
                        CardSampleHelper.create_adaptive_card_submit()
                    )
                )

            elif card_type == CardOptions.hero:
                await step_context.context.send_activity(
                    MessageFactory.attachment(CardSampleHelper.create_hero_card())
                )

            elif card_type == CardOptions.thumbnail:
                await step_context.context.send_activity(
                    MessageFactory.attachment(CardSampleHelper.create_thumbnail_card())
                )

            elif card_type == CardOptions.receipt:
                await step_context.context.send_activity(
                    MessageFactory.attachment(CardSampleHelper.create_receipt_card())
                )

            elif card_type == CardOptions.signin:
                await step_context.context.send_activity(
                    MessageFactory.attachment(CardSampleHelper.create_signin_card())
                )

            elif card_type == CardOptions.carousel:
                #  NOTE if cards are NOT the same height in a carousel,
                #  Teams will instead display as AttachmentLayoutTypes.List
                await step_context.context.send_activity(
                    MessageFactory.carousel(
                        [
                            CardSampleHelper.create_hero_card(),
                            CardSampleHelper.create_hero_card(),
                            CardSampleHelper.create_hero_card(),
                        ]
                    )
                )

            elif card_type == CardOptions.list:
                await step_context.context.send_activity(
                    MessageFactory.list(
                        [
                            CardSampleHelper.create_hero_card(),
                            CardSampleHelper.create_hero_card(),
                            CardSampleHelper.create_hero_card(),
                        ]
                    )
                )

            elif card_type == CardOptions.o365:
                await step_context.context.send_activity(
                    MessageFactory.attachment(
                        CardSampleHelper.create_o365_connector_card()
                    )
                )

            elif card_type == CardOptions.teams_file_consent:
                await step_context.context.send_activity(
                    MessageFactory.attachment(
                        CardSampleHelper.create_teams_file_consent_card(
                            TEAMS_LOGO_FILE_NAME
                        )
                    )
                )

            elif card_type == CardOptions.animation:
                await step_context.context.send_activity(
                    MessageFactory.attachment(
                        CardSampleHelper.create_animation_card(MIND_BLOWN_GIF)
                    )
                )

            elif card_type == CardOptions.audio:
                await step_context.context.send_activity(
                    MessageFactory.attachment(
                        CardSampleHelper.create_audio_card(
                            f"{self.configuration.SERVER_URL}/{MUSIC_API}"
                        )
                    )
                )

            elif card_type == CardOptions.video:
                await step_context.context.send_activity(
                    MessageFactory.attachment(
                        CardSampleHelper.create_video_card(CORGI_ON_CAROUSEL_VIDEO)
                    )
                )

            elif card_type == CardOptions.end:
                # End the dialog so the host gets an EoC
                await step_context.context.send_activity(
                    Activity(
                        type=ActivityTypes.end_of_conversation,
                        code=EndOfConversationCodes.completed_successfully,
                    )
                )
                return DialogTurnResult(DialogTurnStatus.Complete)

        else:
            await step_context.context.send_activity(
                f"{card_type.value} cards are not supported in the {step_context.context.activity.channel_id} channel."
            )

        return await step_context.replace_dialog(
            self.initial_dialog_id, "What card would you want?"
        )

    @staticmethod
    async def card_prompt_validator(prompt_context: PromptValidatorContext) -> bool:
        if not prompt_context.recognized.succeeded:
            if prompt_context.context.activity.attachments:
                return True

            # Render the activity so we can assert in tests.
            # We may need to simplify the json if it gets too complicated to test.
            activity_json = json.dumps(
                prompt_context.context.activity.__dict__, indent=4, default=str
            ).replace("\n", "\r\n")
            prompt_context.options.retry_prompt.text = (
                f"Got {activity_json}\n\n{prompt_context.options.prompt.text}"
            )
            return False
        return True

File: Bots/JavaScript/Skills/CodeFirst/WaterfallSkillBot/dialogs/cards/cardDialog.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { ActionTypes, CardFactory, InputHints, MessageFactory } = require('botbuilder');
const { ComponentDialog, WaterfallDialog, ChoicePrompt, ListStyle, ChoiceFactory, DialogTurnStatus } = require('botbuilder-dialogs');
const fs = require('fs');
const path = require('path');
const { title } = require('process');
const { stringify } = require('querystring');
const { CardOptions } = require('./cardOptions');
const { CardSampleHelper } = require('./cardSampleHelper');
const { ChannelSupportedCards } = require('./channelSupportedCards');

const WATERFALL_DIALOG = 'WaterfallDialog';
const CARD_PROMPT = 'CardPrompt';

class CardDialog extends ComponentDialog {
    /**
     * @param {string} dialogId
     * @param {string} serverUrl
     */
    constructor(dialogId, serverUrl) {
        super(dialogId);
        this.serverUrl = serverUrl;
        this.mindBlownGif = 'https://media3.giphy.com/media/xT0xeJpnrWC4XWblEk/giphy.gif?cid=ecf05e47mye7k75sup6tcmadoom8p1q8u03a7g2p3f76upp9&rid=giphy.gif';
        this.corgiOnCarouselVideo = 'https://www.youtube.com/watch?v=LvqzubPZjHE';
        this.teamsLogoFileName = 'teams-logo.png';

        this.addDialog(new ChoicePrompt(CARD_PROMPT, this.cardPromptValidator))
            .addDialog(new WaterfallDialog(WATERFALL_DIALOG, [
                this.selectCard.bind(this),
                this.displayCard.bind(this)
            ]));

        this.initialDialogId = WATERFALL_DIALOG;
    }

    /**
     * @param {import('botbuilder-dialogs').WaterfallStepContext} stepContext
     */
    async selectCard(stepContext) {
        // Create the PromptOptions from the skill configuration which contain the list of configured skills.
        const messageText = 'What card do you want?';
        const repromptMessageText = 'This message will be created in the validation code';
        const options = {
            prompt: MessageFactory.text(messageText, messageText, InputHints.ExpectingInput),
            retryPrompt: MessageFactory.text(repromptMessageText, repromptMessageText, InputHints.ExpectingInput),
            choices: ChoiceFactory.toChoices(Object.values(CardOptions)),
            style: ListStyle.list
        };

        // Ask the user to enter a card choice.
        return stepContext.prompt(CARD_PROMPT, options);
    }

    /**
     * @param {import('botbuilder-dialogs').WaterfallStepContext} stepContext
     */
    async displayCard(stepContext) {
        if (stepContext.context.activity.value){
            console.log('#####################');
            console.log(stepContext.context.activity);
            await this.handleAdaptiveCard(stepContext);
        }
        else
        {
            const card = stepContext.result.value;
            const cardType = Object.keys(CardOptions).find(key => CardOptions[key].toLowerCase() === card.toLowerCase());
            const { channelId } = stepContext.context.activity;
    
            if (ChannelSupportedCards.isCardSupported(channelId, cardType)) {
                switch (cardType) {
                case CardOptions.AdaptiveCardBotAction:
                    await stepContext.context.sendActivity(MessageFactory.attachment(CardSampleHelper.createAdaptiveCardBotAction()));
                    break;
    
                case CardOptions.AdaptiveCardTaskModule:
                    await stepContext.context.sendActivity(MessageFactory.attachment(CardSampleHelper.createAdaptiveCardTaskModule()));
                    break;
    
                case CardOptions.AdaptiveCardSumbitAction:
                    await stepContext.context.sendActivity(MessageFactory.attachment(CardSampleHelper.createAdaptiveCardSubmit()));
                    break;
    
                case CardOptions.Hero:
                    await stepContext.context.sendActivity(MessageFactory.attachment(CardSampleHelper.createHeroCard()));
                    break;
    
                case CardOptions.Thumbnail:
                    await stepContext.context.sendActivity(MessageFactory.attachment(CardSampleHelper.createThumbnailCard()));
                    break;
    
                case CardOptions.Receipt:
                    await stepContext.context.sendActivity(MessageFactory.attachment(CardSampleHelper.createReceiptCard()));
                    break;
    
                case CardOptions.Signin:
                    await stepContext.context.sendActivity(MessageFactory.attachment(CardSampleHelper.createSigninCard()));
                    break;
    
                case CardOptions.Carousel:
                    // NOTE: if cards are NOT the same height in a carousel, Teams will instead display as AttachmentLayoutTypes.List
                    await stepContext.context.sendActivity(MessageFactory.carousel([
                        CardSampleHelper.createHeroCard(),
                        CardSampleHelper.createHeroCard(),
                        CardSampleHelper.createHeroCard()
                    ]));
                    break;
    
                case CardOptions.List:
                    await stepContext.context.sendActivity(MessageFactory.list([
                        CardSampleHelper.createHeroCard(),
                        CardSampleHelper.createHeroCard(),
                        CardSampleHelper.createHeroCard()
                    ]));
                    break;
    
                case CardOptions.O365:
                    await stepContext.context.sendActivity(MessageFactory.attachment(CardSampleHelper.createO365ConnectorCard()));
                    break;
    
                case CardOptions.TeamsFileConsent: {
                    const file = fs.readFileSync(path.resolve(__dirname, 'files', this.teamsLogoFileName));
                    await stepContext.context.sendActivity(MessageFactory.attachment(CardSampleHelper.createTeamsFileConsentCard(this.teamsLogoFileName, file.byteLength)));
                    break;
                }
    
                case CardOptions.Animation:
                    await stepContext.context.sendActivity(MessageFactory.attachment(CardSampleHelper.createAnimationCard(this.mindBlownGif)));
                    break;
    
                case CardOptions.Audio:
                    await stepContext.context.sendActivity(MessageFactory.attachment(CardSampleHelper.createAudioCard(`${ this.serverUrl }/api/music`)));
                    break;
    
                case CardOptions.Video:
                    await stepContext.context.sendActivity(MessageFactory.attachment(CardSampleHelper.createVideoCard(this.corgiOnCarouselVideo)));
                    break;
                
                case CardOptions.AdaptiveUpdate:
                    var id = await stepContext.context.sendActivity(MessageFactory.attachment(CardDialog.MakeAdaptiveUpdateCard()));
                    console.log(`##### this update card id is ${id.id}`);
                    break;
    
                case CardOptions.End:
                    return { status: DialogTurnStatus.complete };
                }
            } else {
                await stepContext.context.sendActivity(MessageFactory.text(`${ card } cards are not supported in the ${ channelId } channel.`));
            }
        }
        

        return stepContext.replaceDialog(this.initialDialogId);
    }

    static MakeAdaptiveUpdateCard(){
        const { title, text,  buttons } = {
            title: 'Adaptive Update Card',
            text: 'Update Card Action',
            buttons: [
                {
                    type: ActionTypes.MessageBack,
                    title: 'Update card',
                    value: {'count': 0}
                }
            ]
        };

        return CardFactory.heroCard(title, null, buttons, { text });
    }

    /**
     * @param {import('botbuilder-dialogs').WaterfallStepContext} stepContext
     */
    async handleAdaptiveCard(stepContext){
        if (stepContext.context.activity.text === undefined)
        {
            await stepContext.context.sendActivity(MessageFactory.text(`I received an activity with this data in the value field ${JSON.stringify(stepContext.context.activity.value)} `))
        }
        else
        {
            if (stepContext.context.activity.text.includes('update')){
                if (stepContext.context.activity.replyToId){
                    var updatedActivity = MessageFactory.attachment(CardDialog.MakeUpdatedHeroCard(stepContext));
                    console.log(`The bumped card activity id is ${stepContext.context.activity.replyToId}`);
                    updatedActivity.id = stepContext.context.activity.replyToId;
                    await stepContext.context.updateActivity(updatedActivity);
                }
                else
                {
                    await stepContext.context.sendActivity(MessageFactory.text(`Update activity is not supported in the ${stepContext.context.activity.channelId} channel.`));
                }
            }
            else if (stepContext.context.activity.text === ActionTypes.ImBack){
                await stepContext.context.sendActivity(MessageFactory.text("You clicked the imBack button the adaptive card"));
            }
            else
            {
                await stepContext.context.sendActivity(MessageFactory.text(`I received an activity with this data in the text field ${JSON.stringify(stepContext.context.activity.text)} and this data in the value field ${JSON.stringify(stepContext.context.activity.value)}`));
            }
        }
    }

    /**
     * @param {import('botbuilder-dialogs').WaterfallStepContext} stepContext
     */
    static MakeUpdatedHeroCard(stepContext){
        var data = stepContext.context.activity.value;
        data['count'] += 1;

        const { title, text,  buttons } = {
            title: 'Nearly updated card',
            text: `Update count - ${data['count']}`,
            buttons: [
                {
                    type: ActionTypes.MessageBack,
                    title: 'Update card',
                    text: 'Update card text',
                    value: data
                }
            ]
        };

        return CardFactory.heroCard(title, null, buttons, { text });
    }

    /**
     * @param {import('botbuilder-dialogs').PromptValidatorContext<import('botbuilder-dialogs').FoundChoice>} promptContext
     */
    async cardPromptValidator(promptContext) {
        if (!promptContext.recognized.succeeded) {
                // for attachments                             // for adaptive cards                  // for messages from cards
            if (promptContext.context.activity.attachments || promptContext.context.activity.value || promptContext.context.activity.text) {
                return true;
            }

            const activityJson = JSON.stringify(promptContext.context.activity, null, 4).replace(/\n/g, '\r\n');

            // Render the activity so we can assert in tests.
            // We may need to simplify the json if it gets too complicated to test.
            promptContext.options.retryPrompt.text = `Got ${ activityJson }\n\n${ promptContext.options.prompt.text }`;
            return false;
        }

        return true;
    }
}

module.exports.CardDialog = CardDialog;
